doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,298 @@
+	        ...
+	    }
+	    ...
+	}
+
+For the case that multiple PHY ids aggregate in a single syncd context, it is a new separate topic. We don't discuss the topic in this HLD.
+
+The VendorSai overloading constructor will use the functions gbsyncd\_get\_pai\_lib\_name(phy\_id),  gbsyncd\_get\_phy\_access\_lib\_name(phy\_id) and gbsyncd\_get\_mdio\_cl22\_only(phy\_id) to get the JSON key values for "lib\_name", "phy\_access\_lib\_name" and "mdio\_cl22\_only" in the configuration file "gearbox\_config.json".
+
+	#define MDIO_READ            "mdio_read"
+	#define MDIO_WRITE           "mdio_write"
+	#define MDIO_READ_CLAUSE22   "mdio_read_cl22"
+	#define MDIO_WRITE_CLAUSE22  "mdio_write_cl22"
+
+	VendorSai::VendorSai(int phy_id, int context)
+	{
+	    ...
+	    m_context = context;
+	    m_sai_dll_handle = dlopen (gbsyncd_get_pai_lib_name().c_str(), RTLD_LAZY);
+	    *(void**)(&dll_sai_api_initialize) = dlsym(m_sai_dll_handle, "sai_api_initialize");
+	    *(void**)(&dll_sai_api_uninitialize) = dlsym(m_sai_dll_handle, "sai_api_uninitialize");
+	    *(void**)(&dll_sai_api_query) = dlsym(m_sai_dll_handle, "sai_api_query");
+	    ...
+	    m_access_lib_handle = dlopen (gbsyncd_get_phy_access_lib_name().c_str(), RTLD_LAZY);
+	    m_mdio_cl22_only = gbsyncd_get_mdio_cl22_only();
+	    ...
+	    if (m_mdio_cl22_only)
+	    {   
+	        *(void**)(&m_mdio_read) = dlsym(m_access_lib_handle, MDIO_READ_CLAUSE22);
+	    } else {
+	        *(void**)(&m_mdio_read) = dlsym(m_access_lib_handle, MDIO_READ);
+	    }
+	    ...
+	    if (m_mdio_cl22_only)
+	    {   
+	        *(void**)(&m_mdio_write) = dlsym(m_access_lib_handle, MDIO_WRITE_CLAUSE22);
+	    } else {
+	        *(void**)(&m_mdio_write) = dlsym(m_access_lib_handle, MDIO_WRITE);
+	    }   
+	    ...
+	}
+
+Because only the PHY access library name can be changed in the configuration file, we define the PHY access function names always being "mdio\_read", "mdio_write", "mdio\_read\_cl22" and "mdio\_write\_cl22" for MDIO read and write of clause 45 and clause 22, respectively. The access function prototypes are already defined in the PAI library.
+
+The VendorSai::create() function can set the correct the MDIO access functions for PAI in gbsyncd context.
+
+	sai_status_t VendorSai::create(
+        _In_ sai_object_type_t objectType,
+        _Out_ sai_object_id_t* objectId,
+        _In_ sai_object_id_t switchId,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list)
+	{
+	...
+		if (objectType == SAI_OBJECT_TYPE_SWITCH)
+		{
+		    std::vector<sai_attribute_t> attr_copy;
+		    sai_attribute_t attr;
+		    
+		    for (auto i = 0; i < attr_count; i++)
+		    {
+		        if (attr_list[i].id == SAI_SWITCH_ATTR_REGISTER_READ && m_context !=0)
+		        {
+		        attr.id = SAI_SWITCH_ATTR_REGISTER_READ;
+		        attr.value.ptr =  (void *) m_mdio_read;
+		        attr_copy.push_back(attr);
+		        }
+		        else if (attr_list[i].id == SAI_SWITCH_ATTR_REGISTER_WRITE)
```

Comment:
should here be also && m_context !=0 ?

Reply:
Actually SAI_SWITCH_ATTR_REGISTER_WRITE and SAI_SWITCH_ATTR_REGISTER_READ attributes are only passed by orchagent for PAI but not for SAI. As my answer to another comment, we are going to move this piece of code from VendorSai to syncd itself.

Reply:
maybe there should be ParSay inheriting VendorSai to keep this separate from Syncd.cpp class

Reply:
We will create VendorPaiDll inheriting VendorSai. We still need a small change inside VendorSai so that VendorPaiDll does not have to override most the member function.

Reply:
yes